server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,487 @@
+    @Nullable // if the first part hasn't been sent yet
+    private SubscribableListener<ChunkedRestResponseBodyPart> nextAvailableChunksListener;
+
+    /**
+     * A resource to be released when the transmission of the current entry is complete.
+     */
+    @Nullable // if not currently sending an entry
+    private Releasable currentEntryReleasable;
+
+    /**
+     * @param filename     The name of the zip file, which appears in the {@code Content-Disposition} HTTP header of the response, and also
+     *                     is used as a directory prefix for all entries.
+     * @param restChannel  The {@link RestChannel} on which to send the response.
+     * @param onCompletion A resource which is released when the transmission is complete.
+     */
+    public ChunkedZipResponse(String filename, RestChannel restChannel, Releasable onCompletion) {
+        this.filename = filename;
+        this.restChannel = restChannel;
+        this.listenersRefs = AbstractRefCounted.of(() -> enqueueEntry(null, null, ActionListener.releasing(onCompletion)));
+        this.rootListenerRef = Releasables.releaseOnce(listenersRefs::decRef);
+    }
+
+    private final RefCounted listenersRefs;
+    private final Releasable rootListenerRef;
+
+    /**
+     * Close this {@link ChunkedZipResponse}. Once closed, when there are no more pending listeners the zip file footer is sent.
+     */
+    @Override
+    public void close() {
+        rootListenerRef.close();
+    }
+
+    /**
+     * Create a listener which, when completed, will write the result {@link ChunkedRestResponseBodyPart} as an entry in the response stream
+     * with the given name. If the listener is completed successfully with {@code null}, or exceptionally, then no entry is sent. When all
+     * listeners created by this method have been completed, the zip file footer is sent.
+     * <p>
+     * This method may be called as long as this {@link ChunkedZipResponse} is not closed, or there is at least one other incomplete entry
+     * listener.
+     *
+     * @param entryName The name of the entry in the response zip file.
+     * @param listener  A listener which is completed when the entry has been completely processed: either fully sent, or else the request
+     *                  was cancelled and the response will not be used any further. If the returned entry listener is completed
+     *                  exceptionally then the exception is passed to {@code listener}, otherwise this listener is completed successfully.
+     */
+    public ActionListener<ChunkedRestResponseBodyPart> newEntryListener(String entryName, ActionListener<Void> listener) {
+        if (listenersRefs.tryIncRef()) {
+            final var zipEntry = new ZipEntry(filename + "/" + entryName);
+            return ActionListener.assertOnce(ActionListener.releaseAfter(listener.delegateFailureAndWrap((l, firstBodyPart) -> {
+                if (firstBodyPart == null) {
+                    l.onResponse(null);
```

Comment:
I think we will want some way to acquire a listener and then decide not to send its entry. The end-of-file `null` is a purely internal thing, whereas this one is in the public API.

Reply:
Yeah no need to construct an exception just to complete an entry with no content. Remains to be seen tho, I can drop this if you'd prefer as it's easy enough to reinstate if needed later on.

Reply:
I know the end-of-file `null` is an internal thing. But to avoid accidentally leaking this implemention detail, it feels safer to not re-use it for public API. In fact, I'd be OK for either of them to use a different value, e.g. `ChunkedRestResponseBodyPart.END_OF_FILE` or `ChunkedRestResponseBodyPart.IGNORE`, or something similar. So we can keep it using `null` here and use another special value for signal end-of-file?

Reply:
I introduced a `NO_MORE_ENTRIES` sentinel in 0b01e76398e.